### ISSUE TITLE
Key generation is extremely slow

### ISSUE THREAD

**Issue by Loki3000** [phpseclib/phpseclib#2038]
Hello there!

Phpseclib version is 3.0.42

The remote server supports the following allowed algorithms: hmac-sha2-512-etm@openssh.com, hmac-sha2-512, hmac-sha2-256.

Even though I set hmac-sha2-256 for use, key creation is extremely slow:

```
// This executes in about 90 seconds
$curve->multiplyPoint($curve->getBasePoint(), $dA);
```
It uses the nistp521 curve.

How can I speed up key generation?

Thanks a lot!

**Comment by terrafrost** [comment]
I'm not able to reproduce that.

I tried this:

```php
$start = microtime(true);
$private = EC::createKey('nistp521');
$public = $private->getPublicKey();
$elapsed = microtime(true) - $start;

echo $public;
echo "\n";
echo "took $elapsed seconds\n";
```
I tried using various different BigInteger engines as well. The slowest engine (`BigInteger::setEngine('PHP64', ['Reductions\\EvalBarrett'])`) took about 6s, max. And sure, 6s is a long time, too, but (1) it's not 90s and (2) there's a reason phpseclib uses [GMP](https://www.php.net/gmp) if it's available - the GMP extension is sooooo much faster than anything else.

I tried it multiple times and I never got anywhere close to 90s, even with the slowest engine available. I mean, maybe I could if I was at 99% cpu utilization or if my CPU was a 100mhz one but idk.

Regardless, installing the gmp extension, if it's not already installed, should help.
